### ISSUE TITLE
Bug: [no-duplicate-type-constituents] does not handle nested unions/intersections

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#9496]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=C4TwDgpgBAGlC8UB2BXAtgIwgJygHygAoBnYbASyQHN9l0tsBKAbiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type X = number | (string | number);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-duplicate-type-constituents": "error",
  },
};
```

### tsconfig

```jsonc
{}
```

### Expected Result

This code snippet does contain a duplicate type constituent. The rule does report it, but only if the parentheses (which are unnecessary) are removed.

### Actual Result

The duplicate type constituent is not reported by the rule.

### Additional Info

Reported by @auvred in #9479 [(comment)](https://github.com/typescript-eslint/typescript-eslint/pull/9479#discussion_r1665944180).

- Do we care about this issue?
- We'd need to implement a bit of recursion in order to account for this.
- You can only get into this scenario if you have enabled this rule, you are not using ESLint Stylistic (with its `no-extra-parens` rule), and you are not using a formatter (like Prettier).
- Are there similar cases with extraneous parentheses in other rules, and are they handled or not?

**Comment by JoshuaKGoldberg** [comment]
Huh, I don't know why this one took us so long to get to, sorry! Kind of flew under the radar...

> Do we care about this issue?

Yeah, it seems like a straightforward bug to me. Might be tricky and annoying to fix for (_"a bit of recursion"_) but it's a legit case to handle

> Are there similar cases with extraneous parentheses in other rules

Probably 😄 

> and are they handled or not?

Probably ... not. 😂

**Comment by Josh-Cena** [comment]
TIL that `TSUnionType` emits the constituents as an array rather than a binary tree 😅

**Comment by bradzacher** [comment]
It was written that way because that's how TS implements it. Given that union and intersection ordering is commutative there's no need to make it a tree structure - unlike the runtime binary operators which are generally not commutative.

**Comment by bradzacher** [comment]
Wait - how are they not associative?
When taken homogeneously aren't they associative?
What case am I missing that makes them non-associative?

The division `/` and power `**` operators are both neither commutative or associative ever.
But the other math operators are both commutative and associative homogeneously, but are almost never both when heterogeneously.

Logical operators are never commutative. They are associative when taken homogeneously, but are non-associative when taken heterogenously.

Type operators are commutative when taken homogeneously, but not heterogeneously. And are similarly associative when taken homogeneously, but not homogeneously.

... right? Did I get that right?

**Comment by Josh-Cena** [comment]
Commutativity is only defined as $A\oplus B\equiv B\oplus A$. Associativity is only defined as $A\oplus (B\oplus C)\equiv (A\oplus B)\oplus C$ (which I assume you meant by "homogeously"). Only _distributivity_ involves more than one operator.

I would assume, though, that the reason the TS AST works like that is because of things like distributive conditional types.

**Comment by bradzacher** [comment]
![Screenshot_20240827-132600.png of joshes comment in GH mobile](https://github.com/user-attachments/assets/6d4c94bd-82c4-4870-8b91-eb8e71ee60c7)

Big oof the GH mobile app sucks lol. It renders fine in the browser but not in the app. So quality GH.

---

Yeah by homogeneous I meant "same operator" and heterogeneous I meant "different operators".

So then yeah my understanding was correct!

---

> I would assume, though, that the reason the TS AST works like that is because of things like distributive conditional types. 

It would just be an optimisation based on the constraints of the language:

- a small set of operators (eg just unions and intersections)
- the operators are all commutative and assocative
- most commonly the operators are used as long, homogeneous chains

So an array structure makes sense as it allows you to interact with the AST quite easily and quickly. Eg iteration is just an array iteration rather than a tree walk.

For the generic JS binary operators you could specialise certain cases but the reality is that it's just simpler and easier to represent everything as a tree.

**Comment by Josh-Cena** [comment]
Mmm. That makes a lot of sense.
Yeah, I'm aware of the GH mobile funsies. These days I'm on mobile 30% of the time and it really sux to see those unrendered math.

**Comment by JoshuaKGoldberg** [comment]
I'm seeing that comment on desktop and not understanding it even with proper rendering 😛

**Comment by kirkwaiblinger** [comment]
Oh, yeah, I thought there was a way to make the side effects happen in different orders by doing `(a() || b()) || c()` vs `a() || (b() || c())` but I think I'm actually wrong about that. 

Side note - As long as an operation is *associative*, an array of its operands suffices to describe it. If it's commutative as well, a *multiset* would suffice. You only need a binary tree if it's neither 🙂

**Comment by kirkwaiblinger** [comment]
Anyways, back on topic (sort of, at least)....

> Are there similar cases with extraneous parentheses in other rules, and are they handled or not?

There is at least one case of doing this sort of flattening for the runtime logical operators. 

https://github.com/typescript-eslint/typescript-eslint/blob/7160687dd8e270ffb586dcba4eba2c54c63edcd9/packages/eslint-plugin/src/rules/prefer-optional-chain-utils/gatherLogicalOperands.ts#L301-L348

But, as has been noted in previous comments, the AST for this case is different from the type union/intersection operators.
